common/error-with-context/src/lib.rs

```diff
@@ -0,0 +1,130 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+#![forbid(unsafe_code)]
+
+use backtrace::{Backtrace, BacktraceFrame};
+use std::{error::Error, fmt};
+
+/// The ErrorWithContext struct is a wrapper around std::error:Error and provides an
+/// additional struct in which to hold auxiliary information about the context of an error object,
+/// e.g., where the error originated from in the source code, such as the error backtrace. This
+/// is useful for debugging, e.g., when an error is thrown or displayed on the console but the error
+/// message isn't helpful enough to identify exactly what went wrong.
+pub struct ErrorWithContext<T: Error> {
+    pub error: T,
+    pub backtrace: Backtrace,
+}
+
+/// The string displayed in the backtrace frame that shows where the ErrorWithContext object was
+/// initialized. We use this for truncating the backtrace to display only the most helpful backtrace
+/// frames.
+const CREATION_FRAME_STRING: &str = "ErrorWithContext<T>::new";
+
+impl<T: Error> ErrorWithContext<T> {
+    /// Creates and returns an ErrorWithContext struct.
+    pub fn new(error: T) -> ErrorWithContext<T> {
+        // Grab the current backtrace
+        let mut backtrace = Backtrace::new();
+
+        // Truncate the backtrace (if possible) so that when the backtrace is displayed the most
+        // relevant frames are at the top. If we can't find the frame where the ErrorWithContext is
+        // created, fall back to the entire backtrace.
+        let backtrace_frames: Vec<BacktraceFrame> = backtrace.frames().into();
+        if !backtrace_frames.is_empty() {
+            let mut creation_frame_index = 0;
+            for index in 0..backtrace_frames.len() {
+                if let Some(frame) = backtrace_frames.get(index) {
+                    let frame_string = format!("{:?}", frame.symbols());
+                    if frame_string.contains(CREATION_FRAME_STRING) {
+                        creation_frame_index = index;
+                    }
+                }
+            }
+            backtrace = backtrace_frames[creation_frame_index..].to_vec().into();
```

Comment:
then you can eliminate this weirdness

Reply:
also even if you don't want a macro, I wouldn't worry about the extra stuff, though I'd be curious to see what you're clipping here...